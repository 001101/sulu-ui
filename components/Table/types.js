// @flow
type SchemaItem = {
    title: string,
};

export type Schema = {
    [id: string]: SchemaItem,
};

export type Item = {
    id: mixed,
};
