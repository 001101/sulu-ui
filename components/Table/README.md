This is a table, which can be used to present an array of objects. It therefore uses a schema, which has also to be
passed. The schema defines the order of the columns and the title it is displaying in the header.

Mind that every object in the `items` array must contain an `id` value, since it is used to set the
[`key` property of React](https://facebook.github.io/react/docs/lists-and-keys.html#keys).

    const items = [
        {
            id: 1,
            value1: "value 1.1",
            value2: "value 1.2",
        },
        {
            id: 2,
            value1: "value 2.1",
            value2: "value 2.2",
        },
    ];

    const schema = {
        value1: {
            title: "First value",
        },
        value2: {
            title: "Second value",
        },
    };

    <Table items={items} schema={schema}/>
