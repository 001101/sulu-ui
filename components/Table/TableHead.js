// @flow
import React from 'react';

import type {Schema} from './types';

type Props = {
    schema: Schema,
};

export default class TableHead extends React.Component {
    props: Props;

    render() {
        return (
            <thead>
                <tr>
                    {Object.keys(this.props.schema).map((key) => <th key={key}>{this.props.schema[key].title}</th>)}
                </tr>
            </thead>
        );
    }
}
