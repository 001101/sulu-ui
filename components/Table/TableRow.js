// @flow
import React from 'react';

import type {Item, Schema} from './types';

type Props = {
    item: Item,
    schema: Schema,
};

export default class TableRow extends React.Component {
    props: Props;

    render() {
        return (
            <tr>
                {Object.keys(this.props.schema).map((key) => <td key={key}>{this.props.item[key]}</td>)}
            </tr>
        );
    }
}
