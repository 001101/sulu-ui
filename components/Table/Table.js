// @flow
import React from 'react';
import TableHead from './TableHead';
import TableRow from './TableRow';

import type {Item, Schema} from './types';

type Props = {
    items: Array<Item>,
    schema: Schema,
};

export default class Table extends React.Component {
    props: Props;

    render() {
        return (
            <table>
                <TableHead schema={this.props.schema} />
                <tbody>
                    {this.props.items.map((item) => <TableRow key={item.id} item={item} schema={this.props.schema} />)}
                </tbody>
            </table>
        );
    }
}
