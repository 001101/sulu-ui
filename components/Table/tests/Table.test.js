/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import renderer from 'react-test-renderer';
import Table from '../Table';

test('Render data as table', () => {
    const schema = {
        id: {
            title: 'ID',
        },
        title: {
            title: 'title',
        },
        author: {
            title: 'Author',
        },
        authored: {
            title: 'Date',
        },
    };

    const items = [
        {
            id: 1,
            author: 'John Doe',
            authored: '2017-02-28T11:08:41+0100',
            title: 'Test 1',
        },
        {
            id: 2,
            title: 'Test 2',
            author: 'Max Mustermann',
            authored: '2017-02-22T01:08:41+0100',
        },
    ];

    const table = renderer.create(<Table items={items} schema={schema} />);

    expect(table).toMatchSnapshot();
});
