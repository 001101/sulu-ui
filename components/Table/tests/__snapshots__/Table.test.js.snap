// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render data as table 1`] = `
<table>
  <thead>
    <tr>
      <th>
        ID
      </th>
      <th>
        title
      </th>
      <th>
        Author
      </th>
      <th>
        Date
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        1
      </td>
      <td>
        Test 1
      </td>
      <td>
        John Doe
      </td>
      <td>
        2017-02-28T11:08:41+0100
      </td>
    </tr>
    <tr>
      <td>
        2
      </td>
      <td>
        Test 2
      </td>
      <td>
        Max Mustermann
      </td>
      <td>
        2017-02-22T01:08:41+0100
      </td>
    </tr>
  </tbody>
</table>
`;
